Replace network with etcd

ch1bo opened the pull request:
Replaces the existing `Hydra.Node.Network` stack with one only based on `Authenticate` and `Etcd` components. Depicted as in the module docs:

```
-- @
--
--           ▲
--           │                        │
--       Authenticated msg           msg
--           │                        │
--           │                        │
-- ┌─────────┼────────────────────────▼──────┐
-- │                                         │
-- │               Authenticate              │
-- │                                         │
-- └─────────▲────────────────────────┼──────┘
--           │                        │
--           │                        │
--          msg                      msg
--           │                        │
-- ┌─────────┼────────────────────────▼──────┐
-- │                                         │
-- │                   Etcd                  │
-- │                                         │
-- └─────────▲────────────────────────┼──────┘
--           │                        │
--           │        (bytes)         │
--           │                        ▼
--
-- @
```

This is basically a productized version of #1632 which uses a proper grpc client to interact with the `etcd` node (https://hackage.haskell.org/package/grapesy-etcd).

The biggest notable change is the introduction of `NetworkConnected` and `NetworkDisconnected` as the primary indication of whether we are connected to the L2 network or not. Individual peer connectivity is not as important anymore as being connected to the majority cluster is now the deciding factor.

Reliability and fault tolerance of network layer is up to 90% package loss (at least and tested) with tolerance to failing nodes `t < n/2`, while performance seems to be comparable.

~~TBD: Should we also drop `Hydra.Network.Heartbeat` as it is not used anymore?~~ Done this now.

---

* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Haddocks updated
* [x] No new TODOs introduced or explained hereafter

TODO (while reviewing or as follow-up):
- [ ] Change `version` check and determine a network-level compatibility
- [x] Enable `auto-compaction` (done every 5 minutes, maybe use it in the wild instead)
- [x] "Running http and grpc server on single port. This is not recommended for production."

ch1bo committed:
Draft a Hydra.Network.Etcd network component

ch1bo committed:
Define a first test for withEtcdNetwork

ch1bo committed:
Start etcd on network component start

ch1bo committed:
Configure etcd listen host

ch1bo committed:
Re-use NetworkConfiguration for withEtcdNetwork

ch1bo committed:
Configure etcd from node network configuration

ch1bo committed:
Broadcast by putting message using etcdctl

ch1bo committed:
Wait for messages from the etcd cluster

This is naiively using polling for now.

ch1bo committed:
Re-use authentication layer and poll in etcd layer

ch1bo committed:
Do not drop our own messages

ch1bo committed:
Use watch to observe messages on etcd

ch1bo committed:
Use distinct keys per port (hacky) and process multiple updates

ch1bo committed:
Add test to ensure broadcast to self

ch1bo committed:
Hack: disable waitForNodesConnected

The etcd network component does not provide this information (yet)

ch1bo committed:
More deterministic --intial-cluster for etcd and trace stderr

This is quite fragile and requires all --host --port and --peer
arguments to be consistent across the cluster. See the HydraNode.hs
change for example.

ch1bo committed:
Hack: delay startup in case leader election takes some time

This does not work reliably of course. However, we would need to deal
with the situation when we are not connected to the majority anyways.

ch1bo committed:
Add etcd to test nix shells

ch1bo committed:
Disable etcd tests in hydra-node

The last etcd process is not stopping cleanly and I can't get it to
shutdown.

ch1bo committed:
Add etcd to hydra-node docker images

ch1bo committed:
Simplify instructions for fault injection test

Re-using ./seed-devnet (although we don't use all the UTxO) and inlining
the pumba invocation make this easier to copy-paste and run interactively.

ch1bo committed:
Make demo compose file compatible with etcd cluster

The cluster definition is sensitive to inconsistencies between --peer
and --host/--port.

ch1bo committed:
Run more interesting fault injection tests in CI

ch1bo committed:
Update demo process-compose to use etcd

Shellchek chokes on source .env and we can't turn it off in our quite
old pkgs.writeShellApplication. Also the escaping of env variables needs
to be done differently as it seems.

ch1bo committed:
Add tests to ensure behavior when interacting with minority/majority of
etcd cluster

ch1bo committed:
Handle etcdctl watch stopping and reconnect

ch1bo committed:
Handle etcdctl put failing

We expect it to fail when talking to a minority cluster.

However, it is not clear how a failing broadcast should be handled by
the HeadLogic?

ch1bo committed:
Simplify NetworkSpec recordingCallback

ch1bo committed:
Retry broadcasting

This allows a node to not lose messages while it is connected to a
minority of the cluster.

ch1bo committed:
Make etcd network resistant against crash recovery of sender

This uses a basic persistent queue that is backed by files in a
directory. The peek/pop API ensures that files are only removed when the
consumer has processed the event.

ch1bo committed:
Store last known revision on disk

This makes the network component resistant against network partitions
and crashes when the majority carries on broadcasting.

ch1bo committed:
Remove log outputs and delays

None of them seems to be required anymore to make all etcd tests pass!

ch1bo committed:
Step up the fault testing game

Package loss of 10/50/100 percent on randomly selected nodes every 20
seconds for 5 seconds. Should be good values to not trip the timeouts of
the workload (e.g. 20 seconds max for each tx confirmation).

noonio committed:
Example of the simple dropout example on-top of the raft branch

ch1bo committed:
Keep notes on what is still to be done

ch1bo committed:
Replace PeerConnected with NetworkConnected

ch1bo committed:
Add onConnectivity to NetworkCallback

Interleaving 'Connectivity' events with 'msg' of a NetworkCallback
'deliver', is fine until we have a component which provides both (e.g.
the 'Etcd') and one a component that requires symmetry between in- and
out-bound messages. For example the 'Authenticate' component cannot be
composed around such a 'Either Connectivity msg' inbound component
because to verify signatures of incoming messages, it would need to
understand that this is an 'Either' and only do it for 'Right msg' ->
this is a leaky abstraction.

This commit intsead does expand 'NetworkCallback' to not only 'deliver',
but also provide an 'onConnectivity' callback. While this makes it not
a (Co-)functor, it is more explicit and allows to thread callbacks easily.

ch1bo committed:
Add assertion for connectivity events to minority scenario

ch1bo committed:
Decode structured log from etcd

ch1bo committed:
Poll member list for connectivity events in Etcd component

We do not care about the actual member list as it does not indicate who
is connected (but who should be connected in the cluster). But this
query does require us to be on a majority cluster = connectd and will
fail if we are on a minority = disconnected.

ch1bo committed:
Draw network connectivity in TUI

This is now using NetworkConnected instead of PeerConnected

ch1bo committed:
Sketch connectivity events with etcdctl client

This is incomplete as we don't reliable get notified when we are on a
minority cluster. Should switch to a proper etcd client.

ch1bo committed:
Re-enable waitForNodesConnected using NetworkConnected output

ch1bo committed:
Add protocol version to Etcd component and switch withNetwork

This replaces the withNetwork in Hydra.Node.Network now with an
Authenticate + Etcd stack and makes it use the network protocol version.

ch1bo committed:
Encode and check version on etcd keys

This is putting the protocol version into the key name used on the etcd
key value store and checks whether we use the same version on every
watch event. This might be a bit inefficient..

ch1bo committed:
Remove Hydra.Network.{Ouroboros, Reliability}

These are not in use anymore and the Ouroboros/Reliability
implementation is (currently) superseded by the Etcd-based implementation.

ch1bo committed:
Update hackage index

ch1bo committed:
Add grapesy-etcd as dependency

ch1bo committed:
Port Etcd client to using grapesy-etcd

This is not properly handling errors and connectivity topics, but basic
broadcast and delivery works.

ch1bo committed:
Write docs and module haddocks

ch1bo committed:
Trace etcd logs in application tracer

ch1bo committed:
Add a dedicated connectivity test and introduce NetworkConnected

ch1bo committed:
Re-implement peer connectivity (Etcd and Heartbeat)

This is using etcd leases on "alive" keys to determine peer connectivity
on a best-effort basis while ensuring and reporting overall network
connectivity.

ch1bo committed:
Fix ranges and decode alive values defensively

Alive and broadcast messages were interfering because prefixes were
wrong. To get only the keys starting with "msg", we need to use a range
end "msh" as 'h' is 'g'+1.

ch1bo committed:
Bring back PeerConnected and PeerDisconnected

Indicate connectivity to peers by also switching from 'NodeId' to
'Host'. This will make sure that the reported connection/disconnection
is associated with --peer command line arguments.

ch1bo committed:
Add context to GrpcExceptions raised

This should help us locate any errors and complete error handling using
the grapesy grpc client.

ch1bo committed:
Handle and trace failures to broadcast

ch1bo committed:
Not rate limit pings to our etcd node

See comment on 'http2OverridePingRateLimit' for rationale.

ch1bo committed:
Fix treefmt check

ch1bo committed:
Add a static-only variant of snappy dependency

By using that in the hydra-node linking, we always link statically
against this library (could not easily have one derivation yield both,
.a and .so files)

ch1bo committed:
Switch to dockerTools.buildImage

streamLayeredImage was failing with the following error on recent nixpkgs:

 > pyright-root-unknown-node-modules> building node_modules
 > npm error code ENOTCACHED
 > npm error request to https://registry.npmjs.org/yocto-queue failed: cache mode is 'only-if-cached' but no cached response is available.

ch1bo committed:
Use concrete NetworkLog type in HydraLog and fix schema

This is not addressing the TODO in the Network logs schema though.

ch1bo committed:
Not wipe whole persistence dir when restarting in tests

This would make reconnection to the etcd cluster impossible and also
kept a note about this in the changelog.

ch1bo committed:
Create dedicated restart test

This seems to be broken now as the hydra-node process does not exit as
we expect.

locallycompact commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -0,0 +1,513 @@
+        Right x -> pure x
+
+-- | Add context to the 'grpcErrorMessage' of any 'GrpcException' raised.
+withGrpcContext :: MonadCatch m => Text -> m a -> m a
```
I'd prefer to not edit thrown exceptions from a third party library, rather construct new ones (wrapping GrpcException) with more context if needed.

ch1bo replied:
I prefer less code, but will also not fight over it. Feel free to do the change in a follow-up.

locallycompact commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -0,0 +1,513 @@
+      let watchRequest =
+            defMessage
+              & #key .~ "msg"
+              & #rangeEnd .~ "msh" -- NOTE: g+1 to query prefixes
```
Why is this needed?

ch1bo replied:
Which "this" do you mean? What is unclear?

The NOTE on this line says g+1 (and refers to 'h') to get the prefixes.
The NOTE on this variable assignment explains that we "request all keys starting with 'msg'" and gives a link with further details.

locallycompact commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -0,0 +1,513 @@
+                      { broadcast = writePersistentQueue queue
+                      }
+ where
+  connParams doneVar =
```
Could we have local types on this and others? It's hard to see which module this type comes from.

ch1bo replied:
I'd agree if it would not be as straight-forward. Even in low-tech environment like github diff view here shows me the right type and module if I just ctrl+f "connParams".

locallycompact commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -0,0 +1,513 @@
+        & #lease .~ leaseId
+
+  withKeepAlive leaseId action = do
+    biDiStreaming conn (rpc @(Protobuf Lease "leaseKeepAlive")) $ \send _recv ->
```
We experimented with hiding this boilerplate in https://hackage.haskell.org/package/grapesy-etcd-0.3.0/docs/Network-GRPC-Etcd-Client-Simple.html. We could hide all of it if we want. `biDiStreaming`, `NextElem`, and the `rpc` type application could all be encapsulated for all calls like this.

locallycompact commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -0,0 +1,513 @@
+        }
+
+-- * Persistent queue
+
```
`PersistentQueue` looks like it should be in its own module.

ch1bo replied:
Not against that. I will not move it just yet though to not further block this PR.

ch1bo commented on docs/docs/how-to/etcd.md:
Is should drop this before merging

ch1bo committed:
Write specs for restarting nodes

ch1bo committed:
Merge withHydraNode and withHydraNode'

The latter is not actually required and typed-process does not support
variable output stream types (as they are typed).

ch1bo committed:
Fix sub-process handling by issuing SIGINT

Turns out that withHydraNode was directly terminating the hydra-node
sub-process with SIGTERM, which is not handled by the Haskell runtime by
default. As we also have no signal handler, graceful shutdown was not
done.

This is now sending SIGINT from the hydra-cluster tests to the
hydra-node, and from the hydra-node to etcd.

ch1bo committed:
Handle SIGTERM like SIGINT in hydra-node

This makes the hydra-node better behaved by cleaning up
resources (including the etcd sub-process) on receiving SIGTERM, like it
already does for SIGINT.

ch1bo committed:
Fix an e2e tests that was not wiping state fully

If we wipe 'state', we should also clear persisted 'server-output' as
otherwise our clients think the head is already open (while we need to
re-observe txs from chain).

ch1bo committed:
Disable Sanchonet in mithril tests

The network  does not  exist anymore,  we should also  drop it  from the
KnownNetwork enumeration.

ch1bo committed:
Return to using untyped process and checkProcessHasNotDied in withHydraNode

This is able to capture the stderr, which I could not get done with typed-process.

ch1bo committed:
Capture stderr on withHydraNode errors

ch1bo committed:
Capture stderr on withHydraNode errors

ch1bo committed:
Drop redundant test and add known issues about static topology

The test was asserting that the network layer fails to start with wrong
configuration. The etcd-based network is not strictly failing because of
that, but other issue with the cluster can happen which we recorded in
the known issues. Furthermore, there is also the peer
misconfiguration (via the `state` file) which prevents changing the
`--peer` configuration after successful initiation of a head.

ch1bo committed:
Fix connectivity and log decoding errors in Etcd components

This now reports PeerConnected after being disconnectd from the whole
network.

ch1bo committed:
Resolve another todo

Through manual testing, we did not see the watch subscripton fail even
when we try to create it on a minority cluster or when members leave.

ch1bo committed:
Test message resending and enable auto-compaction

We cannot really configure auto-compaction further, but this should
already limit growth of the key value store.

However, this also bounds the number of messages we can miss.

ch1bo committed:
Drop Heartbeat network component

Ths is not used anymore and weeder has been complaining about not-used
code. It might still be useful if we want to use a more "stupid"
transport layer component again, but currently 'Etcd' is providing the
connectivity events itself.

ch1bo committed:
Drop unused definitions

ch1bo committed:
Listen on random free port for etcd HTTP API

This is not actively used, but addresses a warning about
"Running http and grpc server on single port. This is not recommended for production."

noonio commented on hydra-cluster/test/Test/Hydra/Cluster/MithrilSpec.hs:
```diff
@@ -16,6 +17,7 @@ spec :: Spec
 spec = parallel $ do
   describe "downloadLatestSnapshotTo" $
     forEachKnownNetwork "invokes mithril-client correctly" $ \network -> do
+      when (network == Sanchonet) $ pendingWith "Sanchonet not available anymore"
```
Not to do here, but I guess we should just delete sanchoNet entirely now

ch1bo replied:
Yes. I said exactly the same on the commit message of this change.

noonio commented on hydra-node/src/Hydra/Network.hs:
```diff
@@ -122,21 +176,49 @@ readHost s =
-  minPort = fromIntegral (minBound :: Word16)
+-- ** Connectivity & versions
+
+-- TODO: improve these types
```
improve how?

ch1bo replied:
I'll have a follow-up PR for this ready. Currently there is some duplication in `HandshakeFailure` and `HydraHandshakeRefused`, as well as `KnownHydraVersions` actually will not hold multiple versions (we are not using this for switching between multiple implementations, but only to detect incompatibility).

noonio commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -0,0 +1,560 @@
+        , -- Client access only on configured 'host' interface.
+          ["--advertise-client-urls", clientUrl]
+        , -- XXX: Could use unique initial-cluster-tokens to isolate clusters
+          ["--initial-cluster-token", "hydra-network-1"]
```
Doe this mean you can't run multiple clusters on the same computer?

No, I guess, because it's also isloated by the etcd port?

ch1bo replied:
From https://etcd.io/docs/v3.5/op-guide/clustering/

> If spinning up multiple clusters (or creating and destroying a single cluster) with same configuration for testing purpose, it is highly recommended that each cluster is given a unique initial-cluster-token. By doing this, etcd can generate unique cluster IDs and member IDs for the clusters even if they otherwise have the exact same configuration. This can protect etcd from cross-cluster-interaction, which might corrupt the clusters.

noonio commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -0,0 +1,560 @@
+      & #value .~ serialize' msg
+
+  -- TODO: use one key again (after mapping version check)?
+  key = encodeUtf8 @Text $ "msg-" <> show (hydraVersionedProtocolNumber protocolVersion) <> "-" <> show port
```
i thought we wanted to drop this version information in the key?

ch1bo replied:
I'm working on that, but will have it in a separate PR.

noonio commented on hydra-node/src/Hydra/Network/Etcd.hs:
```diff
@@ -0,0 +1,560 @@
+      let watchRequest =
+            defMessage
+              & #key .~ "msg"
+              & #rangeEnd .~ "msh" -- NOTE: g+1 to query prefixes
```
this is hilarious.

noonio commented on hydra-node/src/Hydra/Node/Network.hs:
```diff
@@ -1,211 +1,90 @@
+-- └─────────▲────────────────────────┼──────┘
+--           │                        │
+--           │                        │
+--          msg                      msg
```
```suggestion
--          msg                      Signed msg
```

noonio replied:
right?

ch1bo replied:
Yes, but in both directions. I'll fix it.

ch1bo replied:
I promise I'll include it in the next PR.

noonio commented on hydra-node/test/Hydra/NetworkSpec.hs:
```diff
@@ -8,156 +8,293 @@ import Test.Hydra.Prelude
 
-mockCallback :: Applicative m => NetworkCallback msg m
-mockCallback = NetworkCallback{deliver = \_ -> pure ()}
+noopCallback :: Applicative m => NetworkCallback msg m
```
duplicated in `AuthenticateSpec`

noonio committed:
de-dupe noopCallbac

noonio reviewed (approved):
Looking great :)
